File: Neuron.cpp
#include "Neuron.h"
//#include <cstdlib>
#include <iostream>
#include <time.h>
#include <math.h> 
#include <stdlib.h>

Neuron::Neuron(int inDegree) {
    for(int i = 0; i < inDegree; i++) {
        double w = 0.5; //neurons initialized with arbitrary weight for now. 
        weights.push_back(w);   
    }
}

void Neuron::enterData(layer previousLayer){    
    double sum = 0.0;
    for(int i = 0; i < previousLayer.size(); i++) {
        sum += previousLayer[i].outputValue*weights[i];   
    }
    //do transfer over sum
    this->outputValue = transferFunction(sum);
}

double Neuron::transferFunction(double d) {
    //return (d/(1 + abs(d))); sigmoid
    return tanh(d);
}

void Neuron::calculateOutputDelta(double targetValue) {
    double diff = targetValue - outputValue;
    this->delta = diff*(1 - outputValue*outputValue);
}

void Neuron::calculateHiddenDelta(layer& nextLayer, int index) {
    double sumNextLayer = 0.0;
    for(int i = 0; i < nextLayer.size(); i++) {
        sumNextLayer += nextLayer[i].weights[index]*nextLayer[i].delta;
    }
    //this->delta = (outputValue)*(1-outputValue)*sumNextLayer; this is sigmoid derivative!!!
    this->delta = (1 - outputValue*outputValue)*sumNextLayer;
}

void Neuron::updateWeight(layer& previousLayer) {
    for(int i = 0; i < previousLayer.size(); i++) {
        Neuron& n = previousLayer[i];
        double change = learningRate * delta * n.outputValue;
        this->weights[i] += change;
        
    }
}

double Neuron::learningRate = 0.2; //arbitrary for now. 

File: javaVersion/Neuron.java
import java.util.ArrayList;
import java.util.Random;


public class Neuron {

	private ArrayList<Weight> weights; /* contains weights for connections from all neurons in the previous
	layer as well as the previous change in weight for all those neurons. */ 
	
	public double outputValue; 
	private double delta; //partial of error with respect to weights; used to calculate change in weight.  
	private static double momentum = .3; //how much the old change in weight influences the new change. 
	private static double learningRate = .2; 
	
	public Neuron(int inDegree) {
		weights = new ArrayList<Weight>();
		Random rand = new Random();
	    for(int i = 0; i < inDegree; i++) {
	        weights.add(new Weight(rand.nextDouble(), 0.0));
	    }
	}
	
	void calculateOutput(Layer previousLayer) {    
	    double sum = 0.0;
	    for(int i = 0; i < previousLayer.size(); i++) {
	        sum += previousLayer.getNeuron(i).outputValue*weights.get(i).weight;   
	    }
	    //do transfer over sum
	    outputValue = transferFunction(sum);
	}

	double transferFunction(double d) {
	    return Math.tanh(d);
	}

	void calculateOutputDelta(double targetValue) {
	    double diff = targetValue - outputValue;
	    delta = diff*(1.0 - outputValue*outputValue);
	}

	void calculateHiddenDelta(Layer nextLayer, int index) {
	    double sumNextLayer = 0.0;
	    for(int i = 0; i < nextLayer.size(); i++) {
	        sumNextLayer += nextLayer.getNeuron(i).weights.get(index).weight*nextLayer.getNeuron(i).delta;
	    }
	    
	    delta = (1.0 - outputValue*outputValue)*sumNextLayer;
	}

	void updateWeight(Layer previousLayer) {
		double change;
	    for(int i = 0; i < previousLayer.size(); i++) {
	        Neuron n = previousLayer.getNeuron(i);
	        change = learningRate * delta * n.outputValue + momentum * weights.get(i).previousChangeInWeight;
	        
	        //add change to weight for this neuron, set previousChangeIW to change just computed
	        weights.set(i, new Weight(weights.get(i).weight + change, change)); 
	    }
	    
	}
	
	final class Weight {
		
		public double weight;
		public double previousChangeInWeight;
		
		public Weight(double weight, double previousChange) {
			this.weight = weight;
			this.previousChangeInWeight = previousChange;
		}
	}
	
	
}


